### DISCUSSION TITLE
'TypeError: object with buffer protocol required' when trying to get access token

### DISCUSSION THREAD

**Discussion by PieterHeri** [micropython/micropython#9211] [Using MicroPython]
Hi,

I am trying to get an access token with urequests in MicroPython. This code worked in Python with when using 'requests'. I showed the code and error below. 

Thanks in advance,

```py
import time
import network
import urequests as requests

ssid = 'ExampleWifi'
password = 'Random'

wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.connect(ssid, password)

while wlan.isconnected() == False:
    pass

print('is connected')

u = 'Exampleusername'
p = 'Random'

data = {'grant_type': 'password', 'client_id' : 'anchor',  'username': u, 'password': p}
r = requests.post('https://web.vboxx.nl/oauth/token', data=data)
print('getting access token: {}'.format(r.status_code))
    
dict_str = r.content.decode("UTF-8")`
```

With the following error:

```
is connected
Traceback (most recent call last):
  File "<stdin>", line 21, in <module>
  File "urequests.py", line 184, in post
  File "urequests.py", line 123, in request
TypeError: object with buffer protocol required
```

**Comment by dlech** [comment]
You probably need to convert `data` to a string. A dictionary does not implement the buffer protocol, but strings and bytes do.

**Reply by jimmo** [reply]
FYI @PieterHerings to insert code snippets please use triple-backticks. See https://github.com/orgs/micropython/discussions/9111

To expand more on what @dlech said -- `urequests` only supports specifying raw data (i.e. bytes) as the POST payload, and expects that you have already formatted it in a way that the server understands (and/or matches the specified encoding in the request headers).

The Python `requests` library supports automatically converting a dictionary of k/v pairs into a a form-encoded request.

You can do this in MicroPython by setting the `Content-Type` header explicitly to `application/x-www-form-urlencoded` and then form-encoding the dictionary.

See https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST#example for more info.

Something like:

```
data = ...
data_encoded = '&'.join(n+"="+v for n,v in data)
r = requests.post('https://web.vboxx.nl/oauth/token', headers={"Content-Type": "application/x-www-form-urlencoded"} data=data_encoded)
```

This is something that comes up pretty often so adding support for this in `urequests` might be worth considering.

**Reply by PieterHeri** [reply]
Thanks, I think it works now yet I have another error that I cant find an answer too.
```
payload = { 'grant_type': 'password', 'client_id' : 'anchor', 'username': 'Exampleusername', 'password': 'Random'}
data = ujson.dumps(payload)

r = requests.post('https://web.vboxx.nl/oauth/token', headers={"Content-Type": "application/json"}, data=data).json()
print(r)
```

and I get back: 

``` {'error': 'unsupported_grant_type'} ```

This grant_type works in my Python request code but not in the Micropython urequest code. Do you know why?
